src/extensions/idle.rs

```diff
@@ -77,31 +84,53 @@ impl<'a, T: Read + Write + 'a> Handle<'a, T> {
         unreachable!();
     }
 
-    fn terminate(&mut self) -> Result<()> {
+    fn terminate(&mut self, consume_response: bool) -> Result<()> {
         if !self.done {
             self.done = true;
             self.session.write_line(b"DONE")?;
-            self.session.read_response().map(|_| ())
-        } else {
-            Ok(())
+            if consume_response {
+                return self.session.read_response().map(|_| ());
+            }
         }
+        Ok(())
     }
 
     /// Internal helper that doesn't consume self.
     ///
     /// This is necessary so that we can keep using the inner `Session` in `wait_keepalive`.
     fn wait_inner(&mut self) -> Result<()> {
-        let mut v = Vec::new();
-        match self.session.readline(&mut v).map(|_| ()) {
-            Err(Error::Io(ref e))
-                if e.kind() == io::ErrorKind::TimedOut || e.kind() == io::ErrorKind::WouldBlock =>
-            {
-                // we need to refresh the IDLE connection
-                self.terminate()?;
-                self.init()?;
-                self.wait_inner()
+        let mut buffer = Vec::new();
+        loop {
+            match self.session.readline(&mut buffer).map(|_| ()) {
+                Err(Error::Io(ref e))
+                    if e.kind() == io::ErrorKind::TimedOut
+                        || e.kind() == io::ErrorKind::WouldBlock =>
+                {
+                    if self.keepalive.is_some() {
+                        // we need to refresh the IDLE connection
+                        self.terminate(true)?;
+                        self.init()?;
+                    }
+                }
+                Err(err) => return Err(err),
+                Ok(_) => {
+                    let _ = parse::parse_idle(&buffer, &mut self.unsolicited_responses_tx)?;
```

Comment:
Specifically, shouldn't this use the return value from `parse_idle` to know whether or not `DONE` was sent?